plugins/modules/cloud/misc/terraform.py

```diff
@@ -449,12 +488,70 @@ def main():
     if state == 'present' and module.params.get('parallelism') is not None:
         command.append('-parallelism=%d' % module.params.get('parallelism'))
 
+    def format_args(vars):
+        if isinstance(vars, str):
+            return '"{string}"'.format(string=re.sub('"', '\\"', vars))
+        elif isinstance(vars, bool):
+            if vars:
+                return 'true'
+            else:
+                return 'false'
+        return str(vars)
+
+    def process_complex_args(vars):
+        ret_out = []
+        if isinstance(vars, dict):
+            for k, v in vars.items():
+                if isinstance(v, dict):
+                    ret_out.append('{0}={{{1}}}'.format(k, process_complex_args(v)))
+                elif isinstance(v, list):
+                    ret_out.append("{0}={1}".format(k, process_complex_args(v)))
+                elif isinstance(v, (integer_types, float, str, bool)):
+                    ret_out.append('{0}={1}'.format(k, format_args(v)))
+                else:
+                    # only to handle anything unforeseen
+                    module.fail_json(msg="Supported types are, dictionaries, lists, strings, integer_types, boolean and float.")
+        if isinstance(vars, list):
+            l_out = []
+            for item in vars:
+                if isinstance(item, dict):
+                    l_out.append("{{{0}}}".format(process_complex_args(item)))
+                elif isinstance(item, list):
+                    l_out.append("{0}".format(process_complex_args(item)))
+                elif isinstance(item, (str, integer_types, float, bool)):
+                    l_out.append(format_args(item))
+                else:
+                    # only to handle anything unforeseen
+                    module.fail_json(msg="Supported types are, dictionaries, lists, strings, integer_types, boolean and float.")
+
+            ret_out.append("[{0}]".format(",".join(l_out)))
+        return ",".join(ret_out)
+
     variables_args = []
     for k, v in variables.items():
-        variables_args.extend([
-            '-var',
-            '{0}={1}'.format(k, v)
-        ])
+        if isinstance(v, dict):
+            variables_args.extend([
+                '-var',
+                '{0}={{{1}}}'.format(k, process_complex_args(v))
+            ])
+        elif isinstance(v, list):
+            variables_args.extend([
+                '-var',
+                '{0}={1}'.format(k, process_complex_args(v))
+            ])
+        # terraform does not like shlex_quote fixing double-quotes on the top-level
+        # just passing the plain string and shlex_quote will take care of the rest
```

Comment:
In the last comment thread we established that there is no reason why `shlex_quote` should be mentioned here. Either remove this comment, or adjust it so it is correct.

Reply:
Updated this comment.